source/include/microapp/cs_MicroappStorage.h

```diff
@@ -2,148 +2,143 @@
-
-		/**
-		 * Validate chunk by comparing its calculated checksum with the checksum in the packet.
-		 */
-		uint16_t validateChunk(const uint8_t * const data, uint16_t size, uint16_t compare);
-
-		/**
-		 * Store in flash information about the app (start address, checksum, etc.)
-		 */
-		void storeAppMetadata(uint8_t id, uint16_t checksum, uint16_t size);
-
-
+class MicroappStorage {
+public:
+	static MicroappStorage& getInstance() {
+		static MicroappStorage instance;
+		return instance;
+	}
+
+	/**
+	 * Initialize fstorage, allocate buffer.
+	 */
+	cs_ret_code_t init();
+
+	/**
+	 * Get a copy of the microapp binary header.
+	 *
+	 * @param[in] appIndex   Index of the microapp, validity is not checked.
+	 * @param[out] header    Pointer where will be read to.
+	 */
+	void getAppHeader(uint8_t appIndex, microapp_binary_header_t* header);
+
+	/**
+	 * Get the address in flash, where the microapp program starts.
+	 *
+	 * @param[in] appIndex   Index of the microapp, validity is not checked.
+	 */
+	uint32_t getStartInstructionAddress(uint8_t appIndex);
+
+	/**
+	 * Checks if storage space of this microapp is erased.
+	 *
+	 * @param[in] appIndex   Index of the microapp, validity is not checked.
+	 * @return true                         Storage space of this app is erased, and thus ready to be written.
+	 */
+	bool isErased(uint8_t appIndex);
+
+	/**
+	 * Erases storage space of given app.
+	 * First checks if storage is already erased.
+	 *
+	 * @param[in] appIndex   Index of the microapp, validity is not checked.
+	 *
+	 * @return ERR_SUCCESS_NO_CHANGE        The storage is already erased.
+	 * @return ERR_WAIT_FOR_SUCCESS         The storage will be written erased, wait for EVT_MICROAPP_ERASE_RESULT.
+	 */
+	cs_ret_code_t erase(uint8_t appIndex);
+
+	/**
+	 * Write a chunk to flash.
+	 *
+	 * @param[in] appIndex   Index of the microapp, validity is not checked.
+	 * @param[in] offset     Offset of the data in bytes from the start of the app storage space.
+	 * @param[in] data       Pointer to the data to be written.
+	 * @param[in] size       Size of the data to be written, must be a multiple of 4.
+	 *
+	 * @return ERR_SUCCESS_NO_CHANGE        The data is already on flash.
+	 * @return ERR_WAIT_FOR_SUCCESS         The data will be written to flash, wait for EVT_MICROAPP_UPLOAD_RESULT.
+	 * @return ERR_NO_SPACE                 Data would go outside the app storage space.
+	 * @return ERR_WRONG_PAYLOAD_LENGTH     Data size is not a multiple of 4.
+	 * @return ERR_WRITE_DISABLED           App storage space is not erased.
+	 * @return ERR_BUSY                     Another chunk is being written already.
+	 */
+	cs_ret_code_t writeChunk(uint8_t appIndex, uint16_t offset, const uint8_t* data, uint16_t size);
+
+	/**
+	 * Validate the overall binary, this goes through flash and checks it completely.
+	 * All flash write operations have to have finished before.
+	 *
+	 * @param[in] appIndex   Index of the microapp, validity is not checked.
```

Comment:
how dangerous is it to pass invalid appIndex values? maybe make those functions private?

Reply:
see previous answer

Reply:
Maybe add a `bool isValidIndex(uint8_t appIndex)` method? The class that depends on validity should define when that validity is satisfied to reduce implicit dependency between components (I can't find explicitly which index values are 'valid'). Responsibility of calling this method can be moved to the calling scope.

Reply:
The check is very simple, index < max apps. But i don't see the point of doing the same check over and over again.

Reply:
I was under the impression there was more going on because of possible unused appIndex values. If that is not an issue I would just change the documentation to say `requires appIndex < MAX_APPS` so that it is clear what 'valid' means.